File: quickstart/commands/edge.py
"""qs edge — Build, flash, and monitor the embedded device."""

from __future__ import annotations

import argparse
import os

from quickstart import runner
from quickstart.config import ProjectConfig
from quickstart.env import load_env, make_env
from quickstart.preflight import require


def run_edge(args: argparse.Namespace, config: ProjectConfig) -> int:
    """Flash the embedded firmware and stream serial output. Returns 0 on success."""
    verbose = getattr(args, "verbose", False)
    dry_run = getattr(args, "dry_run", False)

    require(["cargo", "espflash"])

    runner.step("Loading .env.dev")
    env_vars = load_env(config.env_dev)
    env = make_env(env_vars)

    runner.milestone("Building and flashing embedded firmware", emoji="\u26a1 ")
    # .env.dev takes priority, fall back to OS environment
    port = env_vars.get("ESPFLASH_PORT") or os.environ.get("ESPFLASH_PORT", "")
    if port:
        runner.step(f"Using serial port {port}")
    try:
        cmd = ["cargo", "run", "--bin", "supervictor-embedded", "--features", "embedded"]
        if port:
            cmd.extend(["--", "--port", port])
        runner.run(
            cmd,
            cwd=config.device_dir,
            env=env,
            verbose=verbose,
            dry_run=dry_run,
        )
    except Exception:
        runner.error("Flash failed.")
        return 1

    return 0

File: cli/src/commands/edge.rs
use crate::config::ProjectConfig;
use crate::env;
use crate::error::CliError;
use crate::preflight;
use crate::runner::{self, RunOptions, Runner};

/// Arguments for the `qs edge` command.
pub struct EdgeArgs {
    /// Enable verbose output.
    pub verbose: bool,
    /// Print commands without executing.
    pub dry_run: bool,
}

/// Build and flash the ESP32-C3 embedded firmware via espflash.
pub fn run_edge(args: &EdgeArgs, config: &ProjectConfig, r: &dyn Runner) -> Result<i32, CliError> {
    preflight::require(&["cargo"], false, r)?;
    preflight::ensure_espflash(r)?;

    runner::step("Loading .env.dev");
    let env_vars = env::load_env(&config.env_dev)?;
    let env = env::make_env(&env_vars);

    runner::milestone("Building and flashing embedded firmware");

    // .env.dev takes priority, fall back to OS environment
    let port = env_vars
        .get("ESPFLASH_PORT")
        .cloned()
        .or_else(|| std::env::var("ESPFLASH_PORT").ok())
        .unwrap_or_default();

    if !port.is_empty() {
        runner::step(&format!("Using serial port {}", port));
    }

    let mut cmd: Vec<&str> = vec![
        "cargo",
        "run",
        "--bin",
        "supervictor-embedded",
        "--features",
        "embedded",
    ];
    if !port.is_empty() {
        cmd.extend(["--", "--port", &port]);
    }

    match r.run(
        &cmd,
        &RunOptions {
            cwd: Some(config.device_dir.clone()),
            env: Some(env),
            verbose: args.verbose,
            dry_run: args.dry_run,
            ..Default::default()
        },
    ) {
        Ok(_) => Ok(0),
        Err(_) => {
            runner::error("Flash failed.");
            Ok(1)
        }
    }
}

#[cfg(test)]
#[path = "edge_tests.rs"]
mod tests;
